[DI] add tags `container.preload`/`.no_preload` to declare extra classes to preload/services to not preload

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

To allow fine-grained declaration of sidekick classes in DI extensions.
Follows #36103

nicolas-grekas committed:
[DI] add tags `container.preload`/`.no_preload` to declare extra classes to preload/services to not preload

michaljusiega commented:
So `class_exists` is no longer needed in favor of `container.preload` tag do add classes to `*.preload.php` dumped file ?

stof commented:
@michaljusiega AFAIU, the class_exists were solving 2 issues:

- discovery of classes to be preloaded
- compatibility with preloading is some specific cases related to anonymous classes which break if the class is not loaded

This PR provides an alternative solution to the first issue (though `class_exists` will still work). For the second problem, `class_exists` is still the proper solution IMO, as that's not about the Symfony preload dumper by itself, but about compat with preloading.

nicolas-grekas commented:
I updated the PR to now also provide a `container.no_preload` tag.
I added it to eg. console commands.

@michaljusiega: what @stof said: `class_exists()` are better options for portability and for locality. `container.preload` serves adding preloading info to third party packages that don't (want to) use `class_exists()`, and also in programmatic situation where a bundle's logic might known better about fine grained preloading, depending on the configuration.

michaljusiega commented:
@nicolas-grekas, @stof I understand. Thanks!

rosier commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -78,6 +78,7 @@ class PhpDumper extends Dumper
     private $namespace;
     private $asFiles;
     private $hotPathTag;
+    private $preloadTag;
```
```suggestion
    private $preloadTags;
```

fabpot commented:
Thank you @nicolas-grekas.
